ocaml/typing/solver.ml

```diff
@@ -0,0 +1,724 @@
+     guess to replace the constant argument that MIGHT succeed. *)
+  let rec submode_vc :
+      type a. ?log:_ -> a C.obj -> a var -> a -> (unit, a) Result.t =
+    fun (type a) ?log (obj : a C.obj) v a' ->
+     if C.le obj v.upper a'
+     then Ok ()
+     else if not (C.le obj v.lower a')
+     then Error v.lower
+     else (
+       update_upper ?log obj v a';
+       let r =
+         v.vlower
+         |> find_error (fun mu ->
+                let r = submode_mvc ?log obj mu a' in
+                (if Result.is_ok r
+                then
+                  (* Optimization: update [v.lower] based on [mlower u].*)
+                  let mu_lower = mlower obj mu in
+                  if not (C.le obj mu_lower v.lower)
+                  then update_lower ?log obj v mu_lower);
+                r)
+       in
+       if C.le obj v.upper v.lower then set_vlower ?log v [];
+       r)
+
+  and submode_mvc :
+        'a 'r.
+        ?log:change list ref ->
+        'a C.obj ->
+        ('a, allowed * 'r) morphvar ->
+        'a ->
+        (unit, 'a) Result.t =
+   fun ?log obj (Amorphvar (v, f) as mv) a ->
+    (* See [submode_cmv] for why we need the following seemingly redundant
+       lines. *)
+    if C.le obj (mupper obj mv) a
+    then Ok ()
+    else if not (C.le obj (mlower obj mv) a)
+    then Error (mlower obj mv)
+    else
+      let a = C.meet obj a (mupper obj mv) in
+      let f' = C.right_adjoint obj f in
+      let src = C.src obj f in
+      let a' = C.apply src f' a in
+      (* cannot assert [Option.is_none] because [mlower] is imprecise. *)
+      Result.map_error (C.apply obj f) (submode_vc ?log src v a')
+
+  (** Calculate the precise lower bound *)
+  let zap_to_floor_try (type a) (obj : a C.obj) (v : a var) =
+    let rec loop lower =
+      let log = ref [] in
+      let r = submode_vc ~log obj v lower in
+      match r with
+      | Ok () -> log, lower
+      | Error a ->
+        undo_changes !log;
+        loop (C.join obj a lower)
+    in
+    loop v.lower
+
+  let constrain_mlower_try dst (Amorphvar (v, f)) =
+    let src = C.src dst f in
+    let log, lower = zap_to_floor_try src v in
+    log, C.apply dst f lower
+
+  let eq_morphvar :
+      type a l0 r0 l1 r1.
+      a C.obj -> (a, l0 * r0) morphvar -> (a, l1 * r1) morphvar -> bool =
+   fun dst (Amorphvar (v0, f0)) (Amorphvar (v1, f1)) ->
+    let obj0 = C.src dst f0 in
+    let obj1 = C.src dst f1 in
+    match C.eq_obj obj0 obj1 with
+    | None -> false
+    | Some Refl ->
+      v0 == v1
+      && C.disallow_left (C.disallow_right f0)
+         = C.disallow_left (C.disallow_right f1)
+
+  let submode_mvmv (type a) ?log (dst : a C.obj) (Amorphvar (v, f) as mv)
+      (Amorphvar (u, g) as mu) =
+    if C.le dst (mupper dst mv) (mlower dst mu)
+    then Ok ()
+    else if eq_morphvar dst mv mu
+    then Ok ()
+    else
+      (* we have f v <= g u which gives g' (f v) <= u. On the other hand, we
```

Comment:
We're ignoring the "other hand", right? Maybe just drop that bit.

Reply:
The first hand corresponds to the `submode_mvc`, the other hand corresponds to the `submode_cmv`.

Reply:
That's a good observation. Could you put it in the code? In general, I think some commentary of why you need to check both of these conditions would be good.

Reply:
Added comment.